src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs

```diff
@@ -182,17 +269,24 @@ string IDkmClrResultProvider.GetUnderlyingString(DkmEvaluationResult result)
             }
         }
 
-        /// <returns>The qualified name (i.e. including containing types and namespaces) of a named,
-        /// pointer, or array type followed by the qualified name of the actual runtime type, if it
-        /// differs from the declared type.</returns>
-        private static string GetTypeName(DkmInspectionContext inspectionContext, DkmClrType declaredType, DkmClrType runtimeType)
-        {
-            var declaredTypeName = inspectionContext.GetTypeName(declaredType);
+        /// <returns>
+        /// The qualified name (i.e. including containing types and namespaces) of a named, pointer,
+        /// or array type followed by the qualified name of the actual runtime type, if provided.
+        /// </returns>
+        private static string GetTypeName(DkmInspectionContext inspectionContext, DkmClrValue value, DkmClrType declaredType, ExpansionKind kind)
+        { 
             var declaredLmrType = declaredType.GetLmrType();
+            var runtimeType = value.Type;
             var runtimeLmrType = runtimeType.GetLmrType();
-            return declaredLmrType.Equals(runtimeLmrType) || declaredLmrType.IsPointer
-                ? declaredTypeName
-                : string.Format("{0} {{{1}}}", declaredTypeName, inspectionContext.GetTypeName(runtimeType));
+            var includeRuntimeTypeName =
+                !declaredLmrType.Equals(runtimeLmrType) &&
+                !declaredLmrType.IsPointer &&
+                (kind != ExpansionKind.PointerDereference) &&
+                (!declaredLmrType.IsNullable() || value.EvalFlags.Includes(DkmEvaluationResultFlags.ExceptionThrown));
+            var declaredTypeName = inspectionContext.GetTypeName(declaredType, Formatter.NoFormatSpecifiers);
+            return includeRuntimeTypeName ?
+                string.Format("{0} {{{1}}}", declaredTypeName, inspectionContext.GetTypeName(runtimeType, Formatter.NoFormatSpecifiers)) :
```

Comment:
I'm going to leave it for now (but we could certainly do a pass on all the calls to string.Format at some point).